_posts/2017-05-30-database-encryption.markdown

```diff
@@ -0,0 +1,172 @@
+---
+layout: post
+title: "Database Encryption"
+date: 2017-05-30 11:03
+comments: true
+author: ashkan
+categories: [Database, Encryption, Security]
+---
+
+After examining the data stored in one of our systems, we realized it can potentially include some sensitive user data. To reduce the number of people that are technically able to access the data and reduce the risks associated with a potential data theft, we decided to encrypt data getting persisted in the database.
+
+# Our Goal
+Encrypt database without any downtime.
+
+<!-- more -->
+
+# Approach
+While we want to encrypt existing rows, new rows are constantly being added to our database. In order to reach our goal without downtime there was a clear path:
+
+1. Add new _encrypted_ fields to the database.
+2. Start populating new _encrypted_ fields with _encrypted_ values while still populating _non-encrypted_ fields.
+3. Migrate old rows by populating their _encrypted_ fields.
+4. Refactor to use _encrypted_ fields.
+5. Drop _non-encrypted_ fields.
+
+Each step above is relatively simple and can be tested properly before moving to the next step.
+
+## Choosing an Encryption Library
+We looked at few gems, mainly [attr_encrypted](https://github.com/attr-encrypted/attr_encrypted), [crypt_keeper](https://github.com/jmazzi/crypt_keeper) and [symmetric-encryption](https://github.com/rocketjob/symmetric-encryption).
+
+While all these libraries were reliable, we used **Symmetric Encryption**, based on its robust [documentation](https://rocketjob.github.io/symmetric-encryption/), ease of use and easy integration with other libraries (in our case ActiveRecord). It provides some useful [Rake tasks](https://rocketjob.github.io/symmetric-encryption/rake_tasks.html) for configurations inside/outside of Heroku. Overall [`symmetric-encryption`](https://github.com/rocketjob/symmetric-encryption) seemed really 🔒 .
+
+## Configuration
+`config/symmetric-encryption.yml` is used to define what algorithm to use and where to find the related keys for different environments.
+
+Symmetric Encryption uses OpenSSL to encrypt and decrypt the data which means we are able to use any of the algorithms supported by OpenSSL. We used `aes-256-cbc` which is also recommended default algorithm.
+
+In order to create one run:
+
+```bash
+rails generate symmetric_encryption:new_keys production
+```
+
+Generated key above SHOULD NOT be committed into source code. Depending on how your application is deployed, there are two approaches for storing this key. On both scenarios keys are encrypted before storing on file/environment variable. Secret used for encrypting encryption key can be committed into source code.
```

Comment:
_In both scenarios, the keys are encrypted before storage._

It's maybe confusing that we're encrypting the encryption keys themselves. Also counterintuitive that the keys can be committed to the repo. Maybe some background is necessary to explain that we're basically compounding the safeguards by requiring that multiple "factors" be exposed (database _and_ code _and_ config) instead of just the database. Two-factor auth could be a useful analogy.